File: CS/ODataV3Example/ClientForm.cs
using System;
using System.Collections.Generic;
using System.Windows.Forms;
using System.Data.Services.Client;
using System.Linq;
using System.Collections;
using System.IO;
using DevExpress.XtraGrid.Views.Base;
using System.Drawing;
using System.Xml;
using Microsoft.Data.Edm;
using Microsoft.Data.Edm.Csdl;
using Microsoft.Data.Edm.Validation;
using System.Net;
using Microsoft.Data.Edm.Annotations;
using System.ComponentModel;
using DevExpress.Xpo;
using System.Reflection;
using System.Linq.Expressions;
using Microsoft.Data.Edm.Values;

namespace ODataV3Example.Northwind {
    public partial class ClientForm : Form {
        XpoExampleContext context;
        string authCookie = string.Empty;
        readonly Dictionary<string, DataServiceQuery> entities = new Dictionary<string, DataServiceQuery>();
        readonly static Uri serviceUri = new Uri("http://localhost:50293/XpoOdataService.svc/");
        readonly Dictionary<string, Dictionary<string, XpoAnnotation>> annotations = new Dictionary<string, Dictionary<string, XpoAnnotation>>();
        readonly Dictionary<Type, Dictionary<string, Func<object, object>>> properyValuesGetter = new Dictionary<Type, Dictionary<string, Func<object, object>>>();
        IList listObjects;
        IList ListObjects {
            get { return listObjects; }
            set {
                if(listObjects != null) {
                    foreach(INotifyPropertyChanged entity in listObjects)
                        entity.PropertyChanged -= entity_PropertyChanged;
                }
                listObjects = value;
                if(listObjects != null) {
                    foreach(INotifyPropertyChanged entity in listObjects)
                        entity.PropertyChanged += entity_PropertyChanged;
                }
                gclODataV3ServiceMain.DataSource = listObjects;
                gclODataV3ServiceMain.MainView.RefreshData();
                gclODataV3ServiceMain.MainView.PopulateColumns();
            }
        }
        public ClientForm() {
            InitializeComponent();
            CreateContext();
            InitializeEntityList();
            use_Click(null, null);
            GetAnnotation();
        }

        private void InitializeEntityList() {
            entities.Add("Categories", context.Categories);
            entities.Add("Employees", context.Employees);
            entities.Add("Products", context.Products);
            entities.Add("Orders", context.Orders);
            entities.Add("Shippers", context.Shippers);
            entities.Add("Suppliers", context.Suppliers);
            entities.Add("Customers", context.Customers);
            entityToShow.Items.AddRange(entities.Keys.ToArray());
            entityToShow.SelectedIndex = 0;
        }

        private void CreateContext() {
            context = new XpoExampleContext(serviceUri);
            context.SendingRequest += context_SendingRequest;
            context.MergeOption = MergeOption.OverwriteChanges;
        }

        void context_SendingRequest(object sender, SendingRequestEventArgs e) {
            e.RequestHeaders.Add("userCookie", authCookie);
        }

        private void Authorize(object sender, EventArgs e) {
            try {
                UriOperationParameter param1 = new UriOperationParameter("name", "Davolio");
                UriOperationParameter param2 = new UriOperationParameter("password", "abracadabra");
                string res = context.Execute<string>(new Uri("GetAuthorizationCookie", UriKind.Relative), "GET", true, param1, param2).Single();
                authCookie = res;
                entityToShow.Enabled = true;
                entityToShow_SelectedIndexChanged(null, null);
                AuthorizeButton.Enabled = false;
            } catch(Exception ex) {
                MessageBox.Show(ex.Message.ToString());
            }
        }

        private void gvlODataV3Service_FocusedRowChanged(object sender, FocusedRowChangedEventArgs e) {
            pictureBox.Image = null;
            if(ListObjects == null || !(ListObjects[0] is Categories)) return;
            ColumnView view = gclODataV3ServiceMain.FocusedView as ColumnView;
            if(view.Columns.Count < 1 || view.FocusedRowHandle < 0) return;
            object value = view.GetRowCellValue(view.FocusedRowHandle, view.Columns["CategoryID"]);
            if(value == null) {
                pictureBox.Image = null;
            } else {
                int categoryID = (int)value;
                DataServiceStreamResponse resp = context.GetReadStream(context.Categories.Where(i => i.CategoryID == categoryID).Single(), "StreamPicture", new DataServiceRequestArgs());
                pictureBox.Image = Image.FromStream(resp.Stream);
            }
        }

        private void entityToShow_SelectedIndexChanged(object sender, EventArgs e) {
            if(entityToShow.SelectedIndex < 0) return;
            if(string.IsNullOrEmpty(authCookie)) return;
            try {
                context.SaveChanges();
                DataServiceQuery queue = entities[entityToShow.SelectedItem.ToString()];
                List<object> ds = new List<object>();
                ds.AddRange((IEnumerable<object>)queue.Execute());
                ListObjects = ds;
                if(ListObjects[0] is Categories) {
                    categoryInteceptor.Enabled = true;
                    gvlODataV3Service_FocusedRowChanged(null, null);
                } else {
                    categoryInteceptor.Enabled = false;
                    pictureBox.Image = null;
                }
                if(ListObjects[0] is Products) {
                    checkRemove.Enabled = true;
                } else {
                    checkRemove.Enabled = false;
                }
            } catch(Exception ex) {
                MessageBox.Show(ex.ToString());
            }
        }

        private void use_Click(object sender, EventArgs e) {
            UriOperationParameter param1 = new UriOperationParameter("state", true);
            context.Execute(new Uri("SetCategoriesFiltering", UriKind.Relative), "GET", param1);
            entityToShow_SelectedIndexChanged(null, null);
        }

        private void dontUse_Click(object sender, EventArgs e) {
            UriOperationParameter param1 = new UriOperationParameter("state", false);
            context.Execute(new Uri("SetCategoriesFiltering", UriKind.Relative), "GET", param1);
            entityToShow_SelectedIndexChanged(null, null);
        }

        void GetAnnotation() {
            Uri metadataUri = context.GetMetadataUri();
            HttpWebRequest request = (HttpWebRequest)WebRequest.Create(metadataUri);
            HttpWebResponse response = (HttpWebResponse)request.GetResponse();
            Stream stream = response.GetResponseStream();
            StreamReader sr = new StreamReader(stream);
            stream.Flush();
            string metadata = sr.ReadToEnd();

            IEdmModel annotatedModel;
            IEnumerable<EdmError> errors;
            XmlReader xmlReader = XmlReader.Create(new StringReader(metadata));
            bool parsed = EdmxReader.TryParse(xmlReader, out annotatedModel, out errors);

            foreach(IEdmVocabularyAnnotation annotation in annotatedModel.VocabularyAnnotations) {
                if(annotation.Term.TermKind != EdmTermKind.Value) continue;
                IEdmValueAnnotation valueAnnotation = (IEdmValueAnnotation)annotation;
                IEdmProperty property = (IEdmProperty)valueAnnotation.Target;
                IEdmTerm term = valueAnnotation.Term;
                string entityTypeName = ((IEdmSchemaType)property.DeclaringType).Name;
                if(!annotations.ContainsKey(entityTypeName))
                    annotations.Add(entityTypeName, new Dictionary<string, XpoAnnotation>());
                if(!annotations[entityTypeName].ContainsKey(property.Name))
                    annotations[entityTypeName][property.Name] = new XpoAnnotation();
                switch(term.Name) {
                    case "Size":
                        annotations[entityTypeName][property.Name].Size = (int)((IEdmIntegerValue)valueAnnotation.Value).Value;
                        break;
                    case "ReadOnly":
                        annotations[entityTypeName][property.Name].ReadOnly = ((IEdmBooleanValue)valueAnnotation.Value).Value;
                        break;
                }
            }
        }

        void entity_PropertyChanged(object sender, PropertyChangedEventArgs e) {
            Dictionary<string, XpoAnnotation> propertiesAnnotattion;
            if(!annotations.TryGetValue(sender.GetType().Name, out propertiesAnnotattion))
                return;
            XpoAnnotation annotation;
            if(!propertiesAnnotattion.TryGetValue(e.PropertyName, out annotation))
                return;
            if(annotation.ReadOnly) {
                throw new ArgumentException(e.PropertyName);
            }
            object value = GetProperyValue(sender, e.PropertyName);
            if(annotation.Size > 0) {
                if(value.GetType() == typeof(byte[]) && ((byte[])value).Length > annotation.Size)
                    throw new ArgumentException(e.PropertyName);
                if(value.GetType() == typeof(string) && ((string)value).Length > annotation.Size)
                    throw new ArgumentException(e.PropertyName);
            }
        }

        public object GetProperyValue(object entity, string propertyName) {
            if(entity == null) return null;
            Type entityType = entity.GetType();
            Dictionary<string, Func<object, object>> properyGetters;
            if(!properyValuesGetter.TryGetValue(entity.GetType(), out properyGetters))
                properyValuesGetter.Add(entity.GetType(), new Dictionary<string, Func<object, object>>());
            Func<object, object> func;
            if(!properyValuesGetter[entity.GetType()].TryGetValue(propertyName, out func)) {
                Expression body = null;
                ParameterExpression entityParameter = Expression.Parameter(typeof(object), "e");
                Expression instance = Expression.Convert(entityParameter, entityType);
                MemberInfo member = entityType.GetMember(propertyName, BindingFlags.Public | BindingFlags.Instance | BindingFlags.NonPublic).Single();
                body = Expression.MakeMemberAccess(instance, member);
                if(body == null)
                    throw new InvalidOperationException(propertyName);
                if(body.Type != typeof(object)) {
                    body = Expression.Convert(body, typeof(object));
                }
                func = Expression.Lambda<Func<object, object>>(body, entityParameter).Compile();
                properyValuesGetter[entity.GetType()][propertyName] = func;
            }
            return func(entity);
        }

        private void gvlODataV3Service_CellValueChanged(object sender, CellValueChangedEventArgs e) {
            ColumnView view = gclODataV3ServiceMain.FocusedView as ColumnView;
            object entity = view.GetRow(e.RowHandle);
            context.UpdateObject(entity);
        }

        private void checkRemove_Click(object sender, EventArgs e) {
            if(ListObjects == null || !(ListObjects[0] is Products)) return;
            ColumnView view = gclODataV3ServiceMain.FocusedView as ColumnView;
            if(view.Columns.Count < 1 || view.FocusedRowHandle < 0) return;
            int productId = (int)view.GetRowCellValue(view.FocusedRowHandle, view.Columns["ProductID"]);
            Uri actionUri = new Uri(string.Format("{1}({0})/CanRemove", productId, context.Products.RequestUri.AbsoluteUri));
            bool result = context.Execute<bool>(actionUri, "POST", true).Single();
            if(result)
                MessageBox.Show("Can remove");
            else
                MessageBox.Show("Can't remove");
        }
    }

}

File: VB/ODataV3Example/ClientForm.vb
Imports Microsoft.VisualBasic
Imports System
Imports System.Collections.Generic
Imports System.Windows.Forms
Imports System.Data.Services.Client
Imports System.Linq
Imports System.Collections
Imports System.IO
Imports DevExpress.XtraGrid.Views.Base
Imports System.Drawing
Imports System.Xml
Imports Microsoft.Data.Edm
Imports Microsoft.Data.Edm.Csdl
Imports Microsoft.Data.Edm.Validation
Imports System.Net
Imports Microsoft.Data.Edm.Annotations
Imports System.ComponentModel
Imports DevExpress.Xpo
Imports System.Reflection
Imports System.Linq.Expressions
Imports Microsoft.Data.Edm.Values

Namespace ODataV3Example.Northwind
	Partial Public Class ClientForm
		Inherits Form
		Private context As XpoExampleContext
		Private authCookie As String = String.Empty
		Private ReadOnly entities As New Dictionary(Of String, DataServiceQuery)()
		Private ReadOnly Shared serviceUri As New Uri("http://localhost:50293/XpoOdataService.svc/")
		Private ReadOnly annotations As New Dictionary(Of String, Dictionary(Of String, XpoAnnotation))()
		Private ReadOnly properyValuesGetter As New Dictionary(Of Type, Dictionary(Of String, Func(Of Object, Object)))()
		Private listObjects_Renamed As IList
		Private Property ListObjects() As IList
			Get
				Return listObjects_Renamed
			End Get
			Set(ByVal value As IList)
				If listObjects_Renamed IsNot Nothing Then
					For Each entity As INotifyPropertyChanged In listObjects_Renamed
						RemoveHandler entity.PropertyChanged, AddressOf entity_PropertyChanged
					Next entity
				End If
				listObjects_Renamed = value
				If listObjects_Renamed IsNot Nothing Then
					For Each entity As INotifyPropertyChanged In listObjects_Renamed
						AddHandler entity.PropertyChanged, AddressOf entity_PropertyChanged
					Next entity
				End If
				gclODataV3ServiceMain.DataSource = listObjects_Renamed
				gclODataV3ServiceMain.MainView.RefreshData()
				gclODataV3ServiceMain.MainView.PopulateColumns()
			End Set
		End Property
		Public Sub New()
			InitializeComponent()
			CreateContext()
			InitializeEntityList()
			use_Click(Nothing, Nothing)
			GetAnnotation()
		End Sub

		Private Sub InitializeEntityList()
			entities.Add("Categories", context.Categories)
			entities.Add("Employees", context.Employees)
			entities.Add("Products", context.Products)
			entities.Add("Orders", context.Orders)
			entities.Add("Shippers", context.Shippers)
			entities.Add("Suppliers", context.Suppliers)
			entities.Add("Customers", context.Customers)
			entityToShow.Items.AddRange(entities.Keys.ToArray())
			entityToShow.SelectedIndex = 0
		End Sub

		Private Sub CreateContext()
			context = New XpoExampleContext(serviceUri)
			AddHandler context.SendingRequest, AddressOf context_SendingRequest
			context.MergeOption = MergeOption.OverwriteChanges
		End Sub

		Private Sub context_SendingRequest(ByVal sender As Object, ByVal e As SendingRequestEventArgs)
			e.RequestHeaders.Add("userCookie", authCookie)
		End Sub

		Private Sub Authorize(ByVal sender As Object, ByVal e As EventArgs) Handles AuthorizeButton.Click
			Try
				Dim param1 As New UriOperationParameter("name", "Davolio")
				Dim param2 As New UriOperationParameter("password", "abracadabra")
				Dim res As String = context.Execute(Of String)(New Uri("GetAuthorizationCookie", UriKind.Relative), "GET", True, param1, param2).Single()
				authCookie = res
				entityToShow.Enabled = True
				entityToShow_SelectedIndexChanged(Nothing, Nothing)
				AuthorizeButton.Enabled = False
			Catch ex As Exception
				MessageBox.Show(ex.Message.ToString())
			End Try
		End Sub

		Private Sub gvlODataV3Service_FocusedRowChanged(ByVal sender As Object, ByVal e As FocusedRowChangedEventArgs) Handles gvlODataV3Service.FocusedRowChanged
			pictureBox.Image = Nothing
			If ListObjects Is Nothing OrElse Not(TypeOf ListObjects(0) Is Categories) Then
				Return
			End If
			Dim view As ColumnView = TryCast(gclODataV3ServiceMain.FocusedView, ColumnView)
			If view.Columns.Count < 1 OrElse view.FocusedRowHandle < 0 Then
				Return
			End If
			Dim value As Object = view.GetRowCellValue(view.FocusedRowHandle, view.Columns("CategoryID"))
			If value Is Nothing Then
				pictureBox.Image = Nothing
			Else
				Dim categoryID As Integer = CInt(Fix(value))
				Dim resp As DataServiceStreamResponse = context.GetReadStream(context.Categories.Where(Function(i) i.CategoryID = categoryID).Single(), "StreamPicture", New DataServiceRequestArgs())
				pictureBox.Image = Image.FromStream(resp.Stream)
			End If
		End Sub

		Private Sub entityToShow_SelectedIndexChanged(ByVal sender As Object, ByVal e As EventArgs) Handles entityToShow.SelectedIndexChanged
			If entityToShow.SelectedIndex < 0 Then
				Return
			End If
			If String.IsNullOrEmpty(authCookie) Then
				Return
			End If
			Try
				context.SaveChanges()
				Dim queue As DataServiceQuery = entities(entityToShow.SelectedItem.ToString())
				Dim ds As New List(Of Object)()
				ds.AddRange(CType(queue.Execute(), IEnumerable(Of Object)))
				ListObjects = ds
				If TypeOf ListObjects(0) Is Categories Then
					categoryInteceptor.Enabled = True
					gvlODataV3Service_FocusedRowChanged(Nothing, Nothing)
				Else
					categoryInteceptor.Enabled = False
					pictureBox.Image = Nothing
				End If
				If TypeOf ListObjects(0) Is Products Then
					checkRemove.Enabled = True
				Else
					checkRemove.Enabled = False
				End If
			Catch ex As Exception
				MessageBox.Show(ex.ToString())
			End Try
		End Sub

		Private Sub use_Click(ByVal sender As Object, ByVal e As EventArgs) Handles use.Click
			Dim param1 As New UriOperationParameter("state", True)
			context.Execute(New Uri("SetCategoriesFiltering", UriKind.Relative), "GET", param1)
			entityToShow_SelectedIndexChanged(Nothing, Nothing)
		End Sub

		Private Sub dontUse_Click(ByVal sender As Object, ByVal e As EventArgs) Handles dontUse.Click
			Dim param1 As New UriOperationParameter("state", False)
			context.Execute(New Uri("SetCategoriesFiltering", UriKind.Relative), "GET", param1)
			entityToShow_SelectedIndexChanged(Nothing, Nothing)
		End Sub

		Private Sub GetAnnotation()
			Dim metadataUri As Uri = context.GetMetadataUri()
			Dim request As HttpWebRequest = CType(WebRequest.Create(metadataUri), HttpWebRequest)
			Dim response As HttpWebResponse = CType(request.GetResponse(), HttpWebResponse)
			Dim stream As Stream = response.GetResponseStream()
			Dim sr As New StreamReader(stream)
			stream.Flush()
			Dim metadata As String = sr.ReadToEnd()

			Dim annotatedModel As IEdmModel
			Dim errors As IEnumerable(Of EdmError)
			Dim xmlReader As XmlReader = XmlReader.Create(New StringReader(metadata))
			Dim parsed As Boolean = EdmxReader.TryParse(xmlReader, annotatedModel, errors)

			For Each annotation As IEdmVocabularyAnnotation In annotatedModel.VocabularyAnnotations
				If annotation.Term.TermKind <> EdmTermKind.Value Then
					Continue For
				End If
				Dim valueAnnotation As IEdmValueAnnotation = CType(annotation, IEdmValueAnnotation)
				Dim [property] As IEdmProperty = CType(valueAnnotation.Target, IEdmProperty)
				Dim term As IEdmTerm = valueAnnotation.Term
				Dim entityTypeName As String = (CType([property].DeclaringType, IEdmSchemaType)).Name
				If (Not annotations.ContainsKey(entityTypeName)) Then
					annotations.Add(entityTypeName, New Dictionary(Of String, XpoAnnotation)())
				End If
				If (Not annotations(entityTypeName).ContainsKey([property].Name)) Then
					annotations(entityTypeName)([property].Name) = New XpoAnnotation()
				End If
				Select Case term.Name
					Case "Size"
						annotations(entityTypeName)([property].Name).Size = CInt(Fix((CType(valueAnnotation.Value, IEdmIntegerValue)).Value))
					Case "ReadOnly"
						annotations(entityTypeName)([property].Name).ReadOnly = (CType(valueAnnotation.Value, IEdmBooleanValue)).Value
				End Select
			Next annotation
		End Sub

		Private Sub entity_PropertyChanged(ByVal sender As Object, ByVal e As PropertyChangedEventArgs)
			Dim propertiesAnnotattion As Dictionary(Of String, XpoAnnotation)
			If (Not annotations.TryGetValue(sender.GetType().Name, propertiesAnnotattion)) Then
				Return
			End If
			Dim annotation As XpoAnnotation
			If (Not propertiesAnnotattion.TryGetValue(e.PropertyName, annotation)) Then
				Return
			End If
			If annotation.ReadOnly Then
				Throw New ArgumentException(e.PropertyName)
			End If
			Dim value As Object = GetProperyValue(sender, e.PropertyName)
			If annotation.Size > 0 Then
				If value.GetType() Is GetType(Byte()) AndAlso (CType(value, Byte())).Length > annotation.Size Then
					Throw New ArgumentException(e.PropertyName)
				End If
				If value.GetType() Is GetType(String) AndAlso (CStr(value)).Length > annotation.Size Then
					Throw New ArgumentException(e.PropertyName)
				End If
			End If
		End Sub

		Public Function GetProperyValue(ByVal entity As Object, ByVal propertyName As String) As Object
			If entity Is Nothing Then
				Return Nothing
			End If
			Dim entityType As Type = entity.GetType()
			Dim properyGetters As Dictionary(Of String, Func(Of Object, Object))
			If (Not properyValuesGetter.TryGetValue(entity.GetType(), properyGetters)) Then
				properyValuesGetter.Add(entity.GetType(), New Dictionary(Of String, Func(Of Object, Object))())
			End If
			Dim func As Func(Of Object, Object)
			If (Not properyValuesGetter(entity.GetType()).TryGetValue(propertyName, func)) Then
				Dim body As Expression = Nothing
				Dim entityParameter As ParameterExpression = Expression.Parameter(GetType(Object), "e")
				Dim instance As Expression = Expression.Convert(entityParameter, entityType)
				Dim member As MemberInfo = entityType.GetMember(propertyName, BindingFlags.Public Or BindingFlags.Instance Or BindingFlags.NonPublic).Single()
				body = Expression.MakeMemberAccess(instance, member)
				If body Is Nothing Then
					Throw New InvalidOperationException(propertyName)
				End If
				If body.Type IsNot GetType(Object) Then
					body = Expression.Convert(body, GetType(Object))
				End If
				func = Expression.Lambda(Of Func(Of Object, Object))(body, entityParameter).Compile()
				properyValuesGetter(entity.GetType())(propertyName) = func
			End If
			Return func(entity)
		End Function

		Private Sub gvlODataV3Service_CellValueChanged(ByVal sender As Object, ByVal e As CellValueChangedEventArgs) Handles gvlODataV3Service.CellValueChanged
			Dim view As ColumnView = TryCast(gclODataV3ServiceMain.FocusedView, ColumnView)
			Dim entity As Object = view.GetRow(e.RowHandle)
			context.UpdateObject(entity)
		End Sub

		Private Sub checkRemove_Click(ByVal sender As Object, ByVal e As EventArgs) Handles checkRemove.Click
			If ListObjects Is Nothing OrElse Not(TypeOf ListObjects(0) Is Products) Then
				Return
			End If
			Dim view As ColumnView = TryCast(gclODataV3ServiceMain.FocusedView, ColumnView)
			If view.Columns.Count < 1 OrElse view.FocusedRowHandle < 0 Then
				Return
			End If
			Dim productId As Integer = CInt(Fix(view.GetRowCellValue(view.FocusedRowHandle, view.Columns("ProductID"))))
			Dim actionUri As New Uri(String.Format("{1}({0})/CanRemove", productId, context.Products.RequestUri.AbsoluteUri))
			Dim result As Boolean = context.Execute(Of Boolean)(actionUri, "POST", True).Single()
			If result Then
				MessageBox.Show("Can remove")
			Else
				MessageBox.Show("Can't remove")
			End If
		End Sub
	End Class

End Namespace
